python/ray/tune/examples/mnist_tune_serve.py

```diff
@@ -0,0 +1,498 @@
+        x = self.fc(x)
+        x = self.out(x)
+        return F.log_softmax(x, dim=1)
+
+
+def train(model, optimizer, train_loader, device=None):
+    device = device or torch.device("cpu")
+    model.train()
+    for batch_idx, (data, target) in enumerate(train_loader):
+        data, target = data.to(device), target.to(device)
+        optimizer.zero_grad()
+        output = model(data)
+        loss = F.nll_loss(output, target)
+        loss.backward()
+        optimizer.step()
+
+
+def test(model, data_loader, device=None):
+    device = device or torch.device("cpu")
+    model.eval()
+    correct = 0
+    total = 0
+    with torch.no_grad():
+        for batch_idx, (data, target) in enumerate(data_loader):
+            data, target = data.to(device), target.to(device)
+            outputs = model(data)
+            _, predicted = torch.max(outputs.data, 1)
+            total += target.size(0)
+            correct += (predicted == target).sum().item()
+
+    return correct / total
+
+
+# Define Ray Serve model,
+class MNISTBackend:
+    def __init__(self, checkpoint_dir, config, metrics):
+        self.checkpoint_dir = checkpoint_dir
+        self.config = config
+        self.metrics = metrics
+
+        # use_cuda = config.get("use_gpu") and torch.cuda.is_available()
+        # self.device = torch.device("cuda" if use_cuda else "cpu")
+        self.device = "cpu"
+        model = ConvNet(layer_size=self.config["layer_size"]).to(self.device)
+
+        model_state, optimizer_state = torch.load(
+            os.path.join(self.checkpoint_dir, "checkpoint"))
+        model.load_state_dict(model_state)
+
+        self.model = model
+
+    def __call__(self, flask_request):
+        images = torch.tensor(flask_request.json["images"])
+        images = images.to(self.device)
+        outputs = self.model(images)
+        predicted = torch.max(outputs.data, 1)[1]
+        return {"result": predicted.numpy().tolist()}
+
+
+def get_current_model(model_dir):
+    checkpoint_path = os.path.join(model_dir, "checkpoint")
+    meta_path = os.path.join(model_dir, "meta.json")
+
+    if not os.path.exists(checkpoint_path) or \
+       not os.path.exists(meta_path):
+        return None, None, None
+
+    with open(meta_path, "rt") as fp:
+        meta = json.load(fp)
+
+    return checkpoint_path, meta["config"], meta["metrics"]
+
+
+def serve_new_model(model_dir, checkpoint, config, metrics, day):
+    print("Serving checkpoint: {}".format(checkpoint))
+
+    if not os.path.exists(model_dir):
+        try:
+            os.mkdir(model_dir, 0o755)
+        except OSError:
+            return False
+
+    checkpoint_path = os.path.join(model_dir, "checkpoint")
+    meta_path = os.path.join(model_dir, "meta.json")
+
+    if os.path.exists(checkpoint_path):
+        shutil.rmtree(checkpoint_path)
+
+    shutil.copytree(checkpoint, checkpoint_path)
+
+    with open(meta_path, "wt") as fp:
+        json.dump(dict(config=config, metrics=metrics), fp)
```

Comment:
How could I do this on multiple nodes? Maybe leave a comment saying checkpoints could easily be synced to S3 instead